core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java

```diff
@@ -654,6 +683,16 @@ public Builder putMapping(MappingMetaData mappingMd) {
             return this;
         }
 
+        public Builder removeAllPrimaryCandidateAllocationIds() {
+            primaryCandidateAllocationIds.clear();
+            return this;
+        }
+
+        public Builder addPrimaryCandidateAllocationId(String allocationId) {
```

Comment:
I think this will be simpler if we allow setting all the set in one go (see other comment in AllocationService)